python/test/unit/language/test_pipeliner.py

```diff
@@ -33,32 +33,51 @@ def check_capabilities():
 
 @triton.jit
 def matmul_kernel(  #
-        a_ptr, b_ptr, output_ptr,  #
+        a_ptr, scale_ptr, b_ptr, output_ptr,  #
         M, N, K,  #
         stride_am, stride_ak,  #
+        stride_sm, stride_sk,  #
         stride_bk, stride_bn,  #
         stride_cm, stride_cn,  #
         BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,  #
-        NUM_STAGES: tl.constexpr):
+        NUM_STAGES: tl.constexpr, a_type: tl.constexpr, b_type: tl.constexpr):
     pid = tl.program_id(axis=0)
     num_pid_m = tl.cdiv(M, BLOCK_M)
     pid_m = pid % num_pid_m
     pid_n = pid // num_pid_m
     offs_am = (pid_m * BLOCK_M + tl.arange(0, BLOCK_M)) % M
     offs_bn = (pid_n * BLOCK_N + tl.arange(0, BLOCK_N)) % N
+    IS_SCALED: tl.constexpr = a_type is not None and b_type is not None
+    DIV_FACTOR: tl.constexpr = 2 if IS_SCALED and a_type == "e2m1" else 1
+    KA: tl.constexpr = BLOCK_K // DIV_FACTOR
+    BLOCK_AK: tl.constexpr = BLOCK_K // DIV_FACTOR
     offs_k = tl.arange(0, BLOCK_K)
-    a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak)
+    offs_ak = tl.arange(0, BLOCK_AK)
+    a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_ak[None, :] * stride_ak)
     b_ptrs = b_ptr + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+    if IS_SCALED:
+        BLOCK_SK: tl.constexpr = BLOCK_K // 32
+        offs_sk = tl.arange(0, BLOCK_SK)
+        scale_ptrs = scale_ptr + (offs_am[:, None] * stride_sm + offs_sk[None, :] * stride_sk)
     accumulator = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
     for k in tl.range(0, tl.cdiv(K, BLOCK_K), num_stages=NUM_STAGES):
-        mask_a = (offs_am[:, None] < M) & (offs_k[None, :] + k * BLOCK_K < K)
+        mask_a = (offs_am[:, None] < M) & (offs_ak[None, :] + k * BLOCK_AK < KA)
```

Comment:
I wrote this line before as:
```python
mask_a = (offs_am[:, None] < M) & (offs_ak[None, :] + k * BLOCK_AK < K // DIV_FACTOR)
```
and `K // DIV_FACTOR` was not folded into a constant (!!).
This, together with our divisibility analysis for `divi` being rather conservative:
https://github.com/triton-lang/triton/blob/538c237078926a6dd1b4ebb9140d538e1c958d29/lib/Analysis/AxisInfo.cpp#L437-L449
meant that we were not able to prove that `64 / 2` was divisible by anything more than 1 lol.

We should look into improving this constant folding within for-loops.

Reply:
> meant that we were not able to prove that 64 / 2 was divisible by anything more than 1 lol.

I think we can add a new condition where both lhs and rhs are constants

Reply:
nvm, K is not a constant, so we should make explicit that it's a multiple of 16 or 32.

We should probably strengthen our handling of constants in MulOp and DIvOp regardless, but that's a story for another day.